### ISSUE TITLE
feat: Fourier-basis constructive TSP solver

### ISSUE THREAD

**Issue by timgluz** [timgluz/teeline#185]
## Idea

Encode a tour as a closed curve in the plane, represent that curve in Fourier space, and optimize the Fourier coefficients with gradient descent. Decoding always produces a valid permutation via `argsort` of curve parameters — no penalty terms, no validity repair.

This is a **constructive** solver (same category as `nn`), not population-based. No warm-start seeding needed.

---

## Algorithm

**Encoding:** A tour is a closed curve `γ(s) = Σ c_k · e^(2πiks)`, `s ∈ [0,1)`, where `c_k ∈ ℂ` are the degrees of freedom (2K+1 complex coefficients, K << n).

**Energy:**
```
E(c) = Σ_i min_s |city_i − γ(s)|²       # attraction: curve passes near every city
     + λ · Σ_k (2πk)² |c_k|²            # tension: penalizes length (diagonal in Fourier basis!)
```

**Algorithm:**
1. Init: `c_0` = centroid, `c_±1` = small circle, all higher modes = 0
2. Coarse-to-fine loop `K_active = 1..K_max`:
   - Gradient descend `E(c)` w.r.t. active coefficients
   - Attraction gradient: each city pulls its nearest curve point
   - Tension gradient: `−λ(2πk)²c_k` per mode (closed form, no iteration needed)
   - Anneal `λ` downward after each `K_active` stage
3. Decode: for each city, find `s_i = argmin_s |city_i − γ(s)|`; tour = `argsort(s_i)`
4. Optionally pass decoded tour to 2-opt (pipeline: `fourier+2opt`)

**Key property:** The `argsort` decode always produces a valid permutation — unlike Hopfield-Tank networks, there are no penalty terms or validity repair steps.

**Complexity note:** Inner projection loop is O(n·M) per iteration. For berlin52 with M=200 this is manageable. Consider making M scale with n: `M = max(200, 4·n)`.

---

## Python Reference Implementation

Validated: exact optimum on 5 cities on a circle and 8 random uniform cities (0.00% gap both).

```python
import numpy as np
from itertools import permutations

np.random.seed(42)
angles = np.sort(np.random.uniform(0, 2*np.pi, 5))
cities = np.stack([np.cos(angles), np.sin(angles)], axis=1)

def tour_len(perm):
    p = cities[list(perm)]
    return np.sum(np.linalg.norm(p - np.roll(p, -1, axis=0), axis=1))

best = min(permutations(range(5)), key=tour_len)
print(f"optimal tour: {best}  len={tour_len(best):.5f}")

M = 200                      # curve sampling resolution
s = np.arange(M) / M
K = 3
ks = np.arange(-K, K+1)
z = cities[:,0] + 1j*cities[:,1]

c = np.zeros(2*K+1, dtype=complex)
c[K] = z.mean()
c[K+1] = 0.5 * np.mean(np.abs(z - z.mean()))

lam = 0.05
lr = 0.05
basis = np.exp(2j*np.pi*np.outer(ks, s))   # (modes, M)

for K_active in range(1, K+1):
    for it in range(400):
        gamma = c @ basis
        idx = np.abs(z[:,None] - gamma[None,:]).argmin(axis=1)
        grad = np.zeros_like(c)
        for i, j in enumerate(idx):
            err = gamma[j] - z[i]
            grad += err * np.conj(basis[:, j])
        grad += lam * (2*np.pi*ks)**2 * c
        mask = np.abs(ks) <= K_active
        c[mask] -= lr * grad[mask] / len(z)
    lam *= 0.5

gamma = c @ basis
idx = np.abs(z[:,None] - gamma[None,:]).argmin(axis=1)
perm = tuple(int(x) for x in np.argsort(idx))
print(f"fourier tour: {perm}  len={tour_len(perm):.5f}")
```

---

### Options struct (in `src/tsp/mod.rs`, same pattern as `LKOptions`)

```rust
pub struct FourierOptions {
    pub k_max: usize,        // max Fourier mode, default 4
    pub m: usize,            // curve resolution, default 200
    pub lambda: f64,         // initial tension weight, default 0.05
    pub lambda_decay: f64,   // multiplier per K_active stage, default 0.5
    pub lr: f64,             // learning rate, default 0.05
    pub iterations: usize,   // gradient steps per K_active stage, default 400
}
```

### Registration (same four places as every other solver)
1. `Solvers` enum — add `Fourier` variant
2. `FromStr` impl — add `"fourier"` alias
3. `solve_with_context()` dispatch arm
4. `all_meta()` / `SOLVER_LIST` entry (category: Constructive)
5. New file `src/tsp/fourier.rs`

### Complex arithmetic
Use the `num-complex` crate (`num-complex = "0.4"`) for `Complex<f64>`. The basis matrix `e^(2πiks)` and coefficient vector `c` are both `Vec<Complex<f64>>`.

---

## Benchmarks to Run

Priority comparisons on berlin52 with `--optimal-tour` gap metric:

| Pipeline | Expected |
|----------|----------|
| `nn` | +19% (baseline) |
| `fourier` | raw basin quality (unknown — the interesting experiment) |
| `fourier+2opt` | **the target number** |
| `nn+2opt` (`fast` preset) | ~+11% (comparison point) |

Also run a clustered instance (pr76 or similar TSPLIB) — dense clusters are the expected weak point of the projection-decode step.

---

## Related Prior Art

- **Hopfield-Tank (1985):** same spirit (continuous energy landscape), but uses n×n neuron matrix with fragile penalty terms. This formulation avoids both problems.
- **Elastic net (Durbin & Willshaw):** closest relative — also a curve attracted to cities. Difference: elastic net works in spatial coordinates; this works in Fourier coefficients, giving the coarse-to-fine property and a diagonal tension term.
- **Neural TSP solvers (POMO, DIFUSCO):** pretrained, fixed-size, require ONNX — unrelated track.

---

## Acceptance Criteria

- [ ] `src/tsp/fourier.rs` implements the coarse-to-fine gradient descent and `argsort` decode
- [ ] `FourierOptions` struct in `src/tsp/mod.rs` with all six fields, `Default`, `validate()`, `from_toml()`, `from_cli()`
- [ ] Solver registered as `fourier` in enum, `FromStr`, dispatch, and `all_meta()`
- [ ] Unit tests: decode of 5-city circle produces valid permutation; gradient step reduces energy
- [ ] Integration test on berlin52: `fourier` alone produces a valid 52-city tour
- [ ] `docs/benchmarks.md` updated with `fourier` and `fourier+2opt` rows
- [ ] TDD throughout: tests written before implementation
